# Лабораторная работа №1
# Функция Капрекара - K(i,k,r)

### Разделение числа на состовляющие цифры:

In [1]:
def list_from_num(number: int, k: int, r: int):
    list_of_digits = list()
    num=number
    for m in range(k):
        num, tmp = divmod(num, r)
        list_of_digits.append(tmp)
    return list_of_digits

#Тестовые примеры на разные системы счисления
print (list_from_num(1234, 6, 10))
print(list_from_num(10, 4, 2))
print(list_from_num(10,8,2))
print (list_from_num(10, 2, 8))
print(list_from_num(16, 2, 16))
    

[4, 3, 2, 1, 0, 0]
[0, 1, 0, 1]
[0, 1, 0, 1, 0, 0, 0, 0]
[2, 1]
[0, 1]


### Сборка числа из элементов(цифр) списка:

In [2]:
def num_from_list(l:list, k:int, r:int):
    num=0
    for i in range(len(l)):
        num=num*r+l[i];
    return num

#Тестовые примеры на разные системы счисления
print(num_from_list([4, 3, 2, 1, 0, 0], 6,10))
print(num_from_list([1, 1, 0, 1], 4,2))
print(num_from_list([2,1],2,8))
print(num_from_list([1,0],2,16))   
    

432100
13
17
16


### Функция Капрекар:
number - число  
k - разрядность числа  
r - система счисления  

In [3]:
def kaprekar(number:int, k:int, r:int):
    list_of_digits = list_from_num(number,k,r)
    list_of_digits.sort()
    min = num_from_list(list_of_digits, k, r)
    list_of_digits.sort(reverse=True)
    max = num_from_list(list_of_digits, k, r)
    new_list = list(str(max - min))
    new = int("".join(map(str, new_list)))
    return new

print(kaprekar(1000, 4, 10))
print(kaprekar(4196, 4, 10))
print(kaprekar(0, 4, 10))

999
8172
0


In [4]:
def magic_number(num:int, k:int, r:int):
    new_num=kaprekar(num, k, r)
    if new_num == num:
        return num
    else:
        return -1

#Считается ли число магическим?
print(magic_number(6174, 4,10))
print(magic_number(3030,4, 10))
print(magic_number(549945, 6, 10))
    

6174
-1
549945


## Функция, возвращающая список всех магический числел в заданной сиситеме счисления(r), с заданной разрядностью(k):

In [5]:
def magic_numbers(k: int, r: int):
    new = 0
    magic_nums = list()
    exist_nums = list()
    s='9'*k
    for i in range(int(s)+1):
        #print(i, end='')
        while True:
            new = kaprekar(i, k, r)
            #print("---> %d" % new, end='')
            if new == i:
                if new in magic_nums:
                    break
                magic_nums.append(new)
                #print("\n")
                break
            if new in exist_nums:
                #print("--->exist")
                break
            i = new
            exist_nums.append(new)
    return magic_nums
    #print("\nMagic numbers:",magic_nums)  


In [6]:
#Проверка корректной работы функции
print("Magic Numbers:",magic_numbers(4,10))

Magic Numbers: [0, 6174]


## Магические числа, найденные для чисел разной разрядности (k), разных систем счислений (r): 
r=2 k=2 	MAGIC NUMBERS:  [0, 1]  
r=2 k=3 	MAGIC NUMBERS:  [0, 3]  
r=2 k=4 	MAGIC NUMBERS:  [0, 7, 9]  
r=2 k=5 	MAGIC NUMBERS:  [0, 15, 21]  
r=2 k=6 	MAGIC NUMBERS:  [0, 31, 45]  
r=2 k=7 	MAGIC NUMBERS:  [0, 63, 93]  
r=2 k=8 	MAGIC NUMBERS:  [0, 127, 189, 217]  
r=2 k=9 	MAGIC NUMBERS:  [0, 255, 381, 441]  
r=2 k=10 	MAGIC NUMBERS:  [0, 511, 765, 889]  
r=2 k=11 	MAGIC NUMBERS:  [0, 1023, 1533, 1785]  
r=2 k=12 	MAGIC NUMBERS:  [0, 2047, 3069, 3577]  
r=2 k=13 	MAGIC NUMBERS:  [0, 4095, 6141, 7161]  
r=2 k=14 	MAGIC NUMBERS:  [0, 8191, 12285, 14329]  
r=2 k=15 	MAGIC NUMBERS:  [0, 16383, 24573, 28665]  
r=2 k=16 	MAGIC NUMBERS:  [0, 32767, 49149, 57337, 61425]  
r=4 k=2 	MAGIC NUMBERS:  [0]  
r=4 k=3 	MAGIC NUMBERS:  [0, 30]  
r=4 k=4 	MAGIC NUMBERS:  [0, 201]  
r=4 k=5 	MAGIC NUMBERS:  [0]  
r=4 k=6 	MAGIC NUMBERS:  [0, 3369, 2550, 3873]  
r=4 k=7 	MAGIC NUMBERS:  [0, 14565]  
r=4 k=8 	MAGIC NUMBERS:  [0, 54441, 62625]  
r=4 k=9 	MAGIC NUMBERS:  [0, 234405, 171990, 254865]  
r=4 k=10 	MAGIC NUMBERS:  [0, 954261, 873129]  
r=4 k=11 	MAGIC NUMBERS:  [0, 3755685, 4083345]  
r=4 k=12 	MAGIC NUMBERS:  [0, 15285909, 13978281, 16075425]  
r=4 k=13 	MAGIC NUMBERS:  [0, 60111525, 61406805, 65354385, 66665025]  
r=4 k=14 	MAGIC NUMBERS:  [0, 223685289, 244644501, 262470225, 265628289]  
r=4 k=15 	MAGIC NUMBERS:  [0, 961866405, 982776405, 1050933585, 1066723905]  
r=4 k=16 	MAGIC NUMBERS:  [0, 3914590869, 3579095721, 3935304021, 4271106369]  
r=6 k=2 	MAGIC NUMBERS:  [0]  
r=6 k=3 	MAGIC NUMBERS:  [0, 105]  
r=6 k=4 	MAGIC NUMBERS:  [0]  
r=6 k=5 	MAGIC NUMBERS:  [0, 5600]  
r=6 k=6 	MAGIC NUMBERS:  [0, 33860, 42925, 27195]  
r=6 k=7 	MAGIC NUMBERS:  [0]  
r=6 k=8 	MAGIC NUMBERS:  [0, 1275170, 1657225]  
r=6 k=9 	MAGIC NUMBERS:  [0, 6018495]  
r=6 k=10 	MAGIC NUMBERS:  [0, 56319925, 45962330, 47681900]  
r=6 k=11 	MAGIC NUMBERS:  [0, 277695950, 348285175]  
r=6 k=12 	MAGIC NUMBERS:  [0, 2151904825, 1305060855]  
r=6 k=13 	MAGIC NUMBERS:  [0, 10363227560, 12973622725]  
r=6 k=14 	MAGIC NUMBERS:  [0, 73115587525, 60063064790, 59994427550, 78214899025]  
r=6 k=15 	MAGIC NUMBERS:  [0, 451814966575, 469662556825]  
r=6 k=16 	MAGIC NUMBERS:  [0, 2724011240875, 2789618448025]  
r=8 k=2 	MAGIC NUMBERS:  [0, 21]  
r=8 k=3 	MAGIC NUMBERS:  [0, 252]  
r=8 k=4 	MAGIC NUMBERS:  [0]  
r=8 k=5 	MAGIC NUMBERS:  [0]  
r=8 k=6 	MAGIC NUMBERS:  [0, 213402, 147420]  
r=8 k=7 	MAGIC NUMBERS:  [0, 1711962]  
r=8 k=8 	MAGIC NUMBERS:  [0, 16092433]  
r=8 k=9 	MAGIC NUMBERS:  [0, 76545756]  
r=8 k=10 	MAGIC NUMBERS:  [0, 1068263553]  
r=8 k=11 	MAGIC NUMBERS:  [0]  
r=8 k=12 	MAGIC NUMBERS:  [0, 57497839826, 58573445322, 39258683100]  
r=8 k=13 	MAGIC NUMBERS:  [0]  
r=8 k=14 	MAGIC NUMBERS:  [0, 4238480197777, 3749808363210, 3680015145170]  
r=8 k=15 	MAGIC NUMBERS:  [0, 34466323678857, 20104741893852]  
r=8 k=16 	MAGIC NUMBERS:  [0, 280198446204033]  
r=10 k=2 	MAGIC NUMBERS:  [0]  
r=10 k=3 	MAGIC NUMBERS:  [0, 495]  
r=10 k=4 	MAGIC NUMBERS:  [0, 6174]  
r=10 k=5 	MAGIC NUMBERS:  [0]  
r=10 k=6 	MAGIC NUMBERS:  [0, 631764, 549945]  
r=10 k=7 	MAGIC NUMBERS:  [0]  
r=10 k=8 	MAGIC NUMBERS:  [0, 97508421, 63317664]  
r=10 k=9 	MAGIC NUMBERS:  [0, 864197532, 554999445]  
r=10 k=10 	MAGIC NUMBERS:  [0, 9753086421, 9975084201, 6333176664]  
r=10 k=11 	MAGIC NUMBERS:  [0, 86431976532]  
r=10 k=12 	MAGIC NUMBERS:  [0, 975330866421, 555499994445, 997530864201, 999750842001, 633331766664]  
r=10 k=13 	MAGIC NUMBERS:  [0, 8643319766532]  
r=10 k=14 	MAGIC NUMBERS:  [0, 97533308666421, 97755108844221]  
r=10 k=15 	MAGIC NUMBERS:  [0, 864333197666532, 555549999944445]  
r=10 k=16 	MAGIC NUMBERS:  [0, 9775531088644221, 9977551088442201]  
r=12 k=2 	MAGIC NUMBERS:  [0]  
r=12 k=3 	MAGIC NUMBERS:  [0, 858]  
r=12 k=4 	MAGIC NUMBERS:  [0]  
r=12 k=5 	MAGIC NUMBERS:  [0, 172744]  
r=12 k=6 	MAGIC NUMBERS:  [0, 1617330]  
r=12 k=7 	MAGIC NUMBERS:  [0, 28428543]  
r=12 k=8 	MAGIC NUMBERS:  [0, 383468294, 308551012]  
r=12 k=9 	MAGIC NUMBERS:  [0, 2812795986]  
r=12 k=10 	MAGIC NUMBERS:  [0, 44873359960]  
r=12 k=11 	MAGIC NUMBERS:  [0, 663181025078, 534610567348]  
r=12 k=12 	MAGIC NUMBERS:  [0, 8768726887129, 4863092970450]  
r=12 k=13 	MAGIC NUMBERS:  [0, 77728651058344]  
r=12 k=14 	MAGIC NUMBERS:  [0, 1039052057510775, 1048030075321515, 924013465340020, 1282149981815329]  
r=12 k=15 	MAGIC NUMBERS:  [0, 8403796176530898, 11201862503984152]  
r=12 k=16 	MAGIC NUMBERS:  [0, 134342105155721320]  
r=14 k=2 	MAGIC NUMBERS:  [0, 65]  
r=14 k=3 	MAGIC NUMBERS:  [0, 1365]  
r=14 k=4 	MAGIC NUMBERS:  [0]  
r=14 k=5 	MAGIC NUMBERS:  [0]  
r=14 k=6 	MAGIC NUMBERS:  [0, 4033575]  
r=14 k=7 	MAGIC NUMBERS:  [0]  
r=14 k=8 	MAGIC NUMBERS:  [0, 1344056324]  
r=14 k=9 	MAGIC NUMBERS:  [0, 17116239465, 11121123195]  
r=14 k=10 	MAGIC NUMBERS:  [0]  
r=14 k=11 	MAGIC NUMBERS:  [0]  
r=14 k=12 	MAGIC NUMBERS:  [0, 30526696602675]  
r=14 k=13 	MAGIC NUMBERS:  [0, 723266716456440]  
r=14 k=14  

## Функция для вычисления максимальной длины цикла до нахождения магического числа или до зацикливания:

In [9]:
def max_cycle_length(k: int, r: int):
    count1 = 0 
    count2 = 0
    new = 0
    magic_nums = list()
    exist_nums = list()
    for i in range(10000):
        while True:
            new = kaprekar(i, k, r)
            if new == i:
                if new in magic_nums:
                    break
                magic_nums.append(new) 
            elif new in exist_nums:
                break
            i = new
            exist_nums.append(new)            
            count2+=1
        if count2>count1:
            count1 = count2
        count2 = 0
    return count1     

In [19]:
#Перевод из десятичной системы счисления в любую другую
def from_digital(number: int, k: int, r: int):
    list_of_digits = list()
    num = number
    for m in range(k):
        num, tmp = divmod(num, r)
        list_of_digits.append(tmp)
    list_of_digits.reverse()
    l = int("".join(map(str, list_of_digits)))  
    return l
    

In [10]:
for r in range (2, 16 + 1, 2):
    for k in range(2, 16 + 1):
        print("r = ", r, end='')
        print("k = ", k, end='')
        print("\t\t\tMAX_CYCLE_LENGTH: %d" % max_cycle_length(k,r))

r =  2k =  2			MAX_CYCLE_LENGTH: 1
r =  2k =  3			MAX_CYCLE_LENGTH: 2
r =  2k =  4			MAX_CYCLE_LENGTH: 2
r =  2k =  5			MAX_CYCLE_LENGTH: 2
r =  2k =  6			MAX_CYCLE_LENGTH: 2
r =  2k =  7			MAX_CYCLE_LENGTH: 2
r =  2k =  8			MAX_CYCLE_LENGTH: 2
r =  2k =  9			MAX_CYCLE_LENGTH: 2
r =  2k =  10			MAX_CYCLE_LENGTH: 2
r =  2k =  11			MAX_CYCLE_LENGTH: 2
r =  2k =  12			MAX_CYCLE_LENGTH: 2
r =  2k =  13			MAX_CYCLE_LENGTH: 2
r =  2k =  14			MAX_CYCLE_LENGTH: 2
r =  2k =  15			MAX_CYCLE_LENGTH: 2
r =  2k =  16			MAX_CYCLE_LENGTH: 2
r =  4k =  2			MAX_CYCLE_LENGTH: 2
r =  4k =  3			MAX_CYCLE_LENGTH: 4
r =  4k =  4			MAX_CYCLE_LENGTH: 4
r =  4k =  5			MAX_CYCLE_LENGTH: 3
r =  4k =  6			MAX_CYCLE_LENGTH: 7
r =  4k =  7			MAX_CYCLE_LENGTH: 6
r =  4k =  8			MAX_CYCLE_LENGTH: 8
r =  4k =  9			MAX_CYCLE_LENGTH: 6
r =  4k =  10			MAX_CYCLE_LENGTH: 6
r =  4k =  11			MAX_CYCLE_LENGTH: 6
r =  4k =  12			MAX_CYCLE_LENGTH: 7
r =  4k =  13			MAX_CYCLE_LENGTH: 7
r =  4k =  14			MAX_CYCLE_LENGTH: 7
r =  4k 

# Численный эксперемент:
## Задача: 
### Для чисел системы счисления 'r' (от 2 до 10), раздрядности 'k' (от 2 до 6) найти все магические числа и определить максимальную длину цикла

In [94]:
def numerical_experiment():
    statistics = dict()
    statistics = {(0, 0): 0}
    magic = tuple()
    nums = tuple()
    for r in range(2,10+1,1):
        for k in range(2,6+1,1):
            print("r=%d"%r,"k=%d"%k, end='') 
            magic = magic_numbers(k,r)
            print(" \tMAGIC NUMBERS: ", str(magic).ljust(25), end='')
            for i in range(len(magic)):
                magic[i]=from_digital(magic[i],k,r)
            print(' Original number system==({l})'.format(l=magic))
            magic.append(str(max_cycle_length(k,r)))
            statistics[(r, k)]= magic            
    return statistics  

        

In [95]:
numerical_experiment()


r=2 k=2 	MAGIC NUMBERS:  [0, 1]                    Original number system==([0, 1])
r=2 k=3 	MAGIC NUMBERS:  [0, 3]                    Original number system==([0, 11])
r=2 k=4 	MAGIC NUMBERS:  [0, 7, 9]                 Original number system==([0, 111, 1001])
r=2 k=5 	MAGIC NUMBERS:  [0, 15, 21]               Original number system==([0, 1111, 10101])
r=2 k=6 	MAGIC NUMBERS:  [0, 31, 45, 49]           Original number system==([0, 11111, 101101, 110001])
r=3 k=2 	MAGIC NUMBERS:  [0]                       Original number system==([0])
r=3 k=3 	MAGIC NUMBERS:  [0]                       Original number system==([0])
r=3 k=4 	MAGIC NUMBERS:  [0]                       Original number system==([0])
r=3 k=5 	MAGIC NUMBERS:  [0, 184]                  Original number system==([0, 20211])
r=3 k=6 	MAGIC NUMBERS:  [0]                       Original number system==([0])
r=4 k=2 	MAGIC NUMBERS:  [0]                       Original number system==([0])
r=4 k=3 	MAGIC NUMBERS:  [0, 30]                

{(0, 0): 0,
 (2, 2): [0, 1, '1'],
 (2, 3): [0, 11, '2'],
 (2, 4): [0, 111, 1001, '2'],
 (2, 5): [0, 1111, 10101, '2'],
 (2, 6): [0, 11111, 101101, 110001, '2'],
 (3, 2): [0, '2'],
 (3, 3): [0, '2'],
 (3, 4): [0, '3'],
 (3, 5): [0, 20211, '5'],
 (3, 6): [0, '4'],
 (4, 2): [0, '2'],
 (4, 3): [0, 132, '4'],
 (4, 4): [0, 3021, '4'],
 (4, 5): [0, '3'],
 (4, 6): [0, 310221, 213312, 330201, '7'],
 (5, 2): [0, 13, '4'],
 (5, 3): [0, '4'],
 (5, 4): [0, 3032, '4'],
 (5, 5): [0, '6'],
 (5, 6): [0, '7'],
 (6, 2): [0, '3'],
 (6, 3): [0, 253, '5'],
 (6, 4): [0, '8'],
 (6, 5): [0, 41532, '7'],
 (6, 6): [0, 420432, 530421, 325523, '6'],
 (7, 2): [0, '3'],
 (7, 3): [0, '5'],
 (7, 4): [0, '5'],
 (7, 5): [0, '10'],
 (7, 6): [0, '11'],
 (8, 2): [0, 25, '3'],
 (8, 3): [0, 374, '6'],
 (8, 4): [0, '7'],
 (8, 5): [0, '8'],
 (8, 6): [0, 640632, 437734, '10'],
 (9, 2): [0, '4'],
 (9, 3): [0, '6'],
 (9, 4): [0, '5'],
 (9, 5): [0, 62853, '9'],
 (9, 6): [0, '16'],
 (10, 2): [0, '5'],
 (10, 3): [0, 495, '7'],
 (10,

## Значения statistics:  
### формат - (k,r) : [*nums, 'max_length']  
#### k-разрядность  
#### r-система счисления  
#### *nums - список магических точек  
#### max_length - максимальная длина цикла  
{(0, 0): 0,  
 (2, 2): [0, 1, '1'],  
 (2, 3): [0, 11, '2'],  
 (2, 4): [0, 111, 1001, '2'],  
 (2, 5): [0, 1111, 10101, '2'],  
 (2, 6): [0, 11111, 101101, '2'],  
 (3, 2): [0, '2'],  
 (3, 3): [0, '2'],  
 (3, 4): [0, '3'],  
 (3, 5): [0, 20211, '5'],  
 (3, 6): [0, '4'],  
 (4, 2): [0, '2'],  
 (4, 3): [0, 132, '4'],  
 (4, 4): [0, 3021, '4'],  
 (4, 5): [0, '3'],  
 (4, 6): [0, 310221, 213312, 330201, '7'],  
 (5, 2): [0, 13, '4'],  
 (5, 3): [0, '4'],  
 (5, 4): [0, 3032, '4'],  
 (5, 5): [0, '6'],  
 (5, 6): [0, '7'],  
 (6, 2): [0, '3'],  
 (6, 3): [0, 253, '5'],  
 (6, 4): [0, '8'],  
 (6, 5): [0, 41532, '7'],  
 (6, 6): [0, 420432, 530421, 325523, '6'],  
 (7, 2): [0, '3'],  
 (7, 3): [0, '5'],  
 (7, 4): [0, '5'],  
 (7, 5): [0, '10'],  
 (7, 6): [0, '11'],  
 (8, 2): [0, 25, '3'],  
 (8, 3): [0, 374, '6'],  
 (8, 4): [0, '7'],  
 (8, 5): [0, '8'],  
 (8, 6): [0, 640632, 437734, '10'],  
 (9, 2): [0, '4'],  
 (9, 3): [0, '6'],  
 (9, 4): [0, '5'],  
 (9, 5): [0, 62853, '9'],  
 (9, 6): [0, '16'],  
 (10, 2): [0, '5'],  
 (10, 3): [0, 495, '7'],  
 (10, 4): [0, 6174, '6'],  
 (10, 5): [0, '9'],  
 (10, 6): [0, 631764, 549945, '17']}  

# Результаты численного эксперемента
### Полученные данные представлены в таблице
строки таблицы - разрядность числа, колонки - система счисления  
результаты вычислений отображаются в виде [*magic_numbers_list, max_cycle_length]

In [85]:
import pandas as pd
import numpy as np
a = np.zeros((9,5), dtype = list)
s = numerical_experiment()
for key in s:
    a[key[0]-2][key[1]-2]=s[key]
    if key[0] is not 0:
        tmp=a[key[0]-2][key[1]-2][-1]
        del a[key[0]-2][key[1]-2][-1]
        a[key[0]-2][key[1]-2].append("max_cycle = %s"%tmp)      
        
table = pd.DataFrame(a,columns = [str(r) + " digit" for r in range(2, 7)], index = [str(k) + " radix" for k in range(2, 11)])

table

2 digit                  3 digit  \
2 radix    [0, 1, max_cycle = 1]   [0, 11, max_cycle = 2]   
3 radix       [0, max_cycle = 2]       [0, max_cycle = 2]   
4 radix       [0, max_cycle = 2]  [0, 132, max_cycle = 4]   
5 radix   [0, 13, max_cycle = 4]       [0, max_cycle = 4]   
6 radix       [0, max_cycle = 3]  [0, 253, max_cycle = 5]   
7 radix       [0, max_cycle = 3]       [0, max_cycle = 5]   
8 radix   [0, 25, max_cycle = 3]  [0, 374, max_cycle = 6]   
9 radix       [0, max_cycle = 4]       [0, max_cycle = 6]   
10 radix      [0, max_cycle = 5]  [0, 495, max_cycle = 7]   

                                4 digit                          5 digit  \
2 radix   [0, 111, 1001, max_cycle = 2]  [0, 1111, 10101, max_cycle = 2]   
3 radix              [0, max_cycle = 3]        [0, 20211, max_cycle = 5]   
4 radix        [0, 3021, max_cycle = 4]               [0, max_cycle = 3]   
5 radix        [0, 3032, max_cycle = 4]               [0, max_cycle = 6]   
6 radix              [0, max_cycle = 8]        [0, 41532, max_cycle = 7]   
7 radix              [0, max_cycle = 5]              [0, max_cycle = 10]   
8 radix              [0, max_cycle = 7]               [0, max_cycle = 8]   
9 radix              [0, max_cycle = 5]        [0, 62853, max_cycle = 9]   
10 radix       [0, 6174, max_cycle = 6]               [0, max_cycle = 9]   

                                             6 digit  
2 radix            [0, 11111, 101101, max_cycle = 2]  
3 radix                           [0, max_cycle = 4]  
4 radix   [0, 310221, 213312, 330201, max_cycle = 7]  
5 radix                           [0, max_cycle = 7]  
6 radix   [0, 420432, 530421, 325523, max_cycle = 6]  
7 radix                          [0, max_cycle = 11]  
8 radix          [0, 640632, 437734, max_cycle = 10]  
9 radix                          [0, max_cycle = 16]  
10 radix         [0, 631764, 549945, max_cycle = 17]